posthog/api/log_entries.py

```diff
@@ -0,0 +1,121 @@
+from rest_framework.request import Request
+from rest_framework.response import Response
+from rest_framework.decorators import action
+from rest_framework.exceptions import ValidationError
+from rest_framework_dataclasses.serializers import DataclassSerializer
+
+from posthog.clickhouse.client.execute import sync_execute
+
+
+@dataclasses.dataclass(frozen=True)
+class LogEntry:
+    log_source_id: str
+    instance_id: str
+    timestamp: datetime
+    level: str
+    message: str
+
+
+class LogEntrySerializer(DataclassSerializer):
+    class Meta:
+        dataclass = LogEntry
+
+
+class LogEntryRequestSerializer(serializers.Serializer):
+    limit = serializers.IntegerField(required=False)
+    after = serializers.DateTimeField(required=False)
+    before = serializers.DateTimeField(required=False)
+    level = serializers.ListField(child=serializers.CharField(), required=False)
+    search = serializers.CharField(required=False)
+    instance_id = serializers.CharField(required=False)
+
+
+def fetch_log_entries(
+    team_id: int,
+    log_source: str,
+    log_source_id: str,
+    instance_id: Optional[str] = None,
+    after: Optional[datetime] = None,
+    before: Optional[datetime] = None,
+    search: Optional[str] = None,
+    limit: Optional[int] = None,
+    level: Optional[list[str]] = None,
+) -> list[Any]:
+    """Fetch a list of batch export log entries from ClickHouse."""
+    if level is None:
+        level = []
+    clickhouse_where_parts: list[str] = []
+    clickhouse_kwargs: dict[str, Any] = {}
+
+    clickhouse_where_parts.append("log_source = %(log_source)s")
+    clickhouse_kwargs["log_source"] = log_source
+    clickhouse_where_parts.append("log_source_id = %(log_source_id)s")
+    clickhouse_kwargs["log_source_id"] = log_source_id
+    clickhouse_where_parts.append("team_id = %(team_id)s")
+    clickhouse_kwargs["team_id"] = team_id
+
+    if instance_id is not None:
+        clickhouse_where_parts.append("instance_id = %(instance_id)s")
+        clickhouse_kwargs["instance_id"] = instance_id
+    if after is not None:
+        clickhouse_where_parts.append("timestamp > toDateTime64(%(after)s, 6)")
+        clickhouse_kwargs["after"] = after.isoformat().replace("+00:00", "")
+    if before is not None:
+        clickhouse_where_parts.append("timestamp < toDateTime64(%(before)s, 6)")
+        clickhouse_kwargs["before"] = before.isoformat().replace("+00:00", "")
+    if search:
+        clickhouse_where_parts.append("message ILIKE %(search)s")
+        clickhouse_kwargs["search"] = f"%{search}%"
+    if len(level) > 0:
+        clickhouse_where_parts.append("upper(level) in %(levels)s")
+        clickhouse_kwargs["levels"] = level
+
+    clickhouse_query = f"""
+        SELECT log_source_id, instance_id, timestamp, upper(level) as level, message FROM log_entries
+        WHERE {' AND '.join(clickhouse_where_parts)} ORDER BY timestamp DESC {f'LIMIT {limit}' if limit else ''}
```

Comment:
This without a limit (or an Optional limit) feels dangerous

Reply:
Good point. Will update.